nexus/db-queries/src/db/queries/virtual_provisioning_collection_update.rs

```diff
@@ -13,16 +13,52 @@ use crate::db::pool::DbConnection;
 use crate::db::schema::virtual_provisioning_collection;
 use crate::db::schema::virtual_provisioning_resource;
 use crate::db::subquery::{AsQuerySource, Cte, CteBuilder, CteQuery};
+use crate::db::true_or_cast_error::matches_sentinel;
+use crate::db::true_or_cast_error::TrueOrCastError;
 use db_macros::Subquery;
 use diesel::pg::Pg;
 use diesel::query_builder::{AstPass, Query, QueryFragment, QueryId};
+use diesel::result::Error as DieselError;
 use diesel::{
-    sql_types, CombineDsl, ExpressionMethods, IntoSql,
+    sql_types, BoolExpressionMethods, CombineDsl, ExpressionMethods, IntoSql,
     NullableExpressionMethods, QueryDsl, RunQueryDsl, SelectableHelper,
 };
 use nexus_db_model::queries::virtual_provisioning_collection_update::{
-    all_collections, do_update, parent_silo,
+    all_collections, do_update, parent_silo, quotas, silo_provisioned,
 };
+use omicron_common::api::external;
+
+const NOT_ENOUGH_CPUS_SENTINEL: &'static str = "Not enough cpus";
+const NOT_ENOUGH_MEMORY_SENTINEL: &'static str = "Not enough memory";
+const NOT_ENOUGH_STORAGE_SENTINEL: &'static str = "Not enough storage";
+
+/// Translates a generic pool error to an external error based
+/// on messages which may be emitted when provisioning virtual resources
+/// such as instances and disks.
+pub fn from_diesel(e: DieselError) -> external::Error {
+    use crate::db::error;
+
+    let sentinels = [
+        NOT_ENOUGH_CPUS_SENTINEL,
+        NOT_ENOUGH_MEMORY_SENTINEL,
+        NOT_ENOUGH_STORAGE_SENTINEL,
+    ];
+    if let Some(sentinel) = matches_sentinel(&e, &sentinels) {
+        match sentinel {
+            NOT_ENOUGH_CPUS_SENTINEL => {
+                return external::Error::InvalidRequest { message: "Insufficient Capacity: Not enough CPUs to complete request. Either stop unused instances to free up resources or contact the rack operator to request a capacity increase.".to_string() }
+            }
```

Comment:
Eventually it would be nice to use @sunshowers' `InsufficientCapacity` error from https://github.com/oxidecomputer/omicron/pull/4573

Reply:
On second thought, maybe not. I think we should distinguish between hitting up against human set capacity limits vs hardware limits.

Reply:
@zephraph and I chatted about this and post- this is what I suggested:

* Have the `InsufficientCapacity` variant carry around an enum, e.g. `InsufficientCapacityKind::UserQuota` and `::System`, and the `insufficient_capacity` constructor accept that as an argument (forcing callers to specify which `InsufficientCapacityKind` variant they want)
* In the message, include the kind: "Insufficient capacity (user quota)" and "... (system)"
* Consider returning different HTTP error codes for each variant.

cc @askfongjojo who I think has some thoughts about this.

Reply:
@sunshowers @zephraph: Thanks for looping me in. Once silo quotas are fully in place, user should rarely hit the `System` capacity error, assuming that the sum of all silo quotas should not exceed the total usable capacity of the rack. But I can think of two situations when the `System` capacity error would be hit before the `UserQuota` one:
1. Some sleds are in maintenance or taken out of the provisioning pool
2. All sleds are relatively full and the user is trying to provision a very large VM.

As such, perhaps the `UserQuota` kind error will be a HTTP 400 (end user needs to free up some capacity themselves) whereas the `System` one is a 507 (operator needs to migrate some instances, check the sled status, force some silos to reduce their usage temporarily, etc.)?

Reply:
For now I'm just using the `InsufficientCapacity` error wholesale. This isn't the ideal state, it would be good to follow up with a better error state, but it's sufficient for the time being.

Reply:
I created an issue to denote the need for better error messages: https://github.com/oxidecomputer/omicron/issues/4680